File: swift/Tests/ShieldTests/StreamCipherEofTests.swift
import XCTest
@testable import Shield

/// Tests for the authenticated end-of-stream tag in StreamCipher.
final class StreamCipherEofTests: XCTestCase {

    // Cross-language golden vector:
    //   master_key = 32 x 0x42, stream_salt = 16 x 0x01, chunk_count = 3
    private let expectedTagHex =
        "52d4dfbeccc364bd69a2f232aa460bd1eb79b0c93903f344dd7b937703918431"

    private func hex(_ b: [UInt8]) -> String {
        b.map { String(format: "%02x", $0) }.joined()
    }

    /// Recomputes the end-of-stream tag using the same byte layout the
    /// StreamCipher uses, and checks it against the shared golden vector.
    func testEofTagConformanceVector() {
        let masterKey = [UInt8](repeating: 0x42, count: 32)
        let streamSalt = [UInt8](repeating: 0x01, count: 16)
        let chunkCount: UInt64 = 3

        let eofKey = Shield.hmacSha256(key: masterKey, data: Array("shield-stream-eof".utf8))
        var input = streamSalt
        input.append(contentsOf: withUnsafeBytes(of: chunkCount.littleEndian) { Array($0) })
        let tag = Shield.hmacSha256(key: eofKey, data: input)

        XCTAssertEqual(hex(tag), expectedTagHex)
    }

    func testStreamRoundtrip() throws {
        let key = [UInt8](repeating: 0x42, count: 32)
        let cipher = try StreamCipher(key: key, chunkSize: 16)
        let data = (0..<64).map { UInt8($0) }
        XCTAssertEqual(try cipher.decrypt(cipher.encrypt(data)), data)
    }

    func testTruncationAtChunkBoundaryRejected() throws {
        let key = [UInt8](repeating: 0x42, count: 32)
        let cipher = try StreamCipher(key: key, chunkSize: 16)
        let data = (0..<64).map { UInt8($0) }
        let enc = try cipher.encrypt(data)
        // Drop the trailing 4-byte zero marker and the 32-byte EOF tag.
        let truncated = Array(enc[0..<(enc.count - 36)])
        XCTAssertThrowsError(try cipher.decrypt(truncated))
    }

    func testForgedEndMarkerRejected() throws {
        let key = [UInt8](repeating: 0x42, count: 32)
        let cipher = try StreamCipher(key: key, chunkSize: 16)
        let data = (0..<64).map { UInt8($0) }
        let enc = try cipher.encrypt(data)
        // Strip trailer, then re-append a bare zero marker (no valid tag).
        var forged = Array(enc[0..<(enc.count - 36)])
        forged.append(contentsOf: [0, 0, 0, 0])
        XCTAssertThrowsError(try cipher.decrypt(forged))
    }
}

File: csharp/Shield.Tests/StreamCipherEofTests.cs
using System;
using System.Linq;
using System.Reflection;
using System.Security.Cryptography;
using Xunit;
using Dikestra.Shield;

namespace Shield.Tests
{
    public class StreamCipherEofTests
    {
        // Cross-language golden vector for the authenticated end-of-stream tag.
        //   master_key  = 32 x 0x42
        //   stream_salt = 16 x 0x01
        //   chunk_count = 3
        private const string ExpectedTagHex =
            "52d4dfbeccc364bd69a2f232aa460bd1eb79b0c93903f344dd7b937703918431";

        private static byte[] ComputeEofTag(byte[] key, byte[] salt, ulong count)
        {
            var m = typeof(StreamCipher).GetMethod(
                "ComputeEofTag", BindingFlags.NonPublic | BindingFlags.Static);
            return (byte[])m.Invoke(null, new object[] { key, salt, count });
        }

        private static string ToHex(byte[] b) =>
            string.Concat(b.Select(x => x.ToString("x2")));

        [Fact]
        public void TestEofTagConformanceVector()
        {
            byte[] key = Enumerable.Repeat((byte)0x42, 32).ToArray();
            byte[] salt = Enumerable.Repeat((byte)0x01, 16).ToArray();
            byte[] tag = ComputeEofTag(key, salt, 3);
            Assert.Equal(ExpectedTagHex, ToHex(tag));
        }

        [Fact]
        public void TestStreamRoundtrip()
        {
            byte[] key = Enumerable.Repeat((byte)0x42, 32).ToArray();
            using var cipher = new StreamCipher(key, 16);
            byte[] data = Enumerable.Range(0, 64).Select(i => (byte)i).ToArray();
            byte[] enc = cipher.Encrypt(data);
            Assert.Equal(data, cipher.Decrypt(enc));
        }

        [Fact]
        public void TestTruncationAtChunkBoundaryRejected()
        {
            byte[] key = Enumerable.Repeat((byte)0x42, 32).ToArray();
            using var cipher = new StreamCipher(key, 16);
            byte[] data = Enumerable.Range(0, 64).Select(i => (byte)i).ToArray();
            byte[] enc = cipher.Encrypt(data);

            // Drop the trailing 32-byte EOF tag and the 4-byte zero marker.
            byte[] truncated = enc.Take(enc.Length - 36).ToArray();
            Assert.ThrowsAny<Exception>(() => cipher.Decrypt(truncated));
        }

        [Fact]
        public void TestForgedEndMarkerRejected()
        {
            byte[] key = Enumerable.Repeat((byte)0x42, 32).ToArray();
            using var cipher = new StreamCipher(key, 16);
            byte[] data = Enumerable.Range(0, 64).Select(i => (byte)i).ToArray();
            byte[] enc = cipher.Encrypt(data);

            // Strip trailer, then re-append a bare zero marker (no valid tag).
            byte[] forged = enc.Take(enc.Length - 36)
                .Concat(new byte[] { 0, 0, 0, 0 }).ToArray();
            Assert.ThrowsAny<Exception>(() => cipher.Decrypt(forged));
        }
    }
}
